components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go

```diff
@@ -0,0 +1,328 @@
+			utils.NewConfigValidationFieldRequiredError(path, "i2c_addr"))
+	})
+}
+
+func TestConnect(t *testing.T) {
+	logger := golog.NewTestLogger(t)
+	ctx := context.Background()
+	cancelCtx, cancelFunc := context.WithCancel(ctx)
+	g := rtkI2C{
+		cancelCtx:  cancelCtx,
+		cancelFunc: cancelFunc,
+		logger:     logger,
+	}
+
+	url := "http://fakeurl"
+	username := "user"
+	password := "pwd"
+
+	// create new ntrip client and connect
+	err := g.connect("invalidurl", username, password, 10)
+	test.That(t, err.Error(), test.ShouldContainSubstring, `Can't connect to NTRIP caster`)
+
+	g.ntripClient = makeMockNtripClient()
+
+	err = g.connect(url, username, password, 10)
+	test.That(t, err, test.ShouldBeNil)
+
+	err = g.getStream("", 10)
+	test.That(t, err.Error(), test.ShouldContainSubstring, `lookup fakeurl`)
+}
+
+func TestReadings(t *testing.T) {
+	var (
+		alt   = 50.5
+		speed = 5.4
+		loc   = geo.NewPoint(40.7, -73.98)
+	)
+
+	logger := golog.NewTestLogger(t)
+	ctx := context.Background()
+	cancelCtx, cancelFunc := context.WithCancel(ctx)
+	g := rtkI2C{
+		cancelCtx:  cancelCtx,
+		cancelFunc: cancelFunc,
+		logger:     logger,
+	}
+
+	mockSensor := &CustomMovementSensor{
+		MovementSensor: &fake.MovementSensor{},
+	}
+
+	mockSensor.PositionFunc = func(ctx context.Context, extra map[string]interface{}) (*geo.Point, float64, error) {
+		return loc, alt, nil
+	}
+
+	g.nmeamovementsensor = mockSensor
+
+	// Normal position
+	loc1, alt1, err := g.Position(ctx, make(map[string]interface{}))
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, loc1, test.ShouldResemble, loc)
+	test.That(t, alt1, test.ShouldEqual, alt)
+
+	speed1, err := g.LinearVelocity(ctx, make(map[string]interface{}))
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, speed1.Y, test.ShouldEqual, speed)
+	test.That(t, speed1.X, test.ShouldEqual, 0)
+	test.That(t, speed1.Z, test.ShouldEqual, 0)
+
+	// Zero position with latitude 0 and longitude 0.
+	mockSensor.PositionFunc = func(ctx context.Context, extra map[string]interface{}) (*geo.Point, float64, error) {
+		return geo.NewPoint(0, 0), 0, nil
+	}
+
+	loc2, alt2, err := g.Position(ctx, make(map[string]interface{}))
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, loc2, test.ShouldResemble, geo.NewPoint(0, 0))
+	test.That(t, alt2, test.ShouldEqual, 0)
+
+	speed2, err := g.LinearVelocity(ctx, make(map[string]interface{}))
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, speed2.Y, test.ShouldEqual, speed)
+	test.That(t, speed2.X, test.ShouldEqual, 0)
+	test.That(t, speed2.Z, test.ShouldEqual, 0)
+
+	// Position with NaN values.
+	mockSensor.PositionFunc = func(ctx context.Context, extra map[string]interface{}) (*geo.Point, float64, error) {
+		return geo.NewPoint(math.NaN(), math.NaN()), math.NaN(), nil
+	}
+
+	g.lastposition.SetLastPosition(loc1)
+
+	loc3, alt3, err := g.Position(ctx, make(map[string]interface{}))
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, loc3, test.ShouldResemble, loc1)
```

Comment:
was a little confused when I first read this line, a comment explaining that it should be returning the last position would be helpful!